lib/src/main/java/com/diffplug/spotless/generic/PipeStepPair.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2020 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.generic;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.diffplug.spotless.FormatterStep;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+public class PipeStepPair {
+	/** The two steps will be named `<name>In` and `<name>Out`. */
+	public static Builder named(String name) {
+		return new Builder(name);
+	}
+
+	public static String defaultToggleName() {
+		return "toggle";
+	}
+
+	public static String defaultToggleOff() {
+		return "spotless:off";
+	}
+
+	public static String defaultToggleOn() {
+		return "spotless:on";
+	}
```

Comment:
I'm feeling ambivalent about these two methods.

On one hand, it feels like `PipeStepPair` should keep this information internal, because otherwise it could cause unnecessary code coupling.

But on the other hand, having a quick look at these methods' usages (see `FormatExtension.toggleOffOn()`, `ToggleOffOn.off` and `ToggleOffOn.on`), it's not obvious to me what we could do to remove these methods without introducing lots more code.

Thoughts?

Reply:
I had the same instinct, and got stuck in the same place.  I take comfort in the "ugliness" with the same point about "overspecialization" above.  If we use this PipeStep to implement other "special rules within this block" sorts of constructs, then I think we'll be glad that we kept the API minimal, and left the `spotless:off`-specific bits as just default strings, rather than baking them into the API.